# h6 Hello Django

## Tehtävä a

Käytän tehtävän teossa Teron tekemää ohjetta joka löytyy [täältä](https://terokarvinen.com/2022/django-instant-crm-tutorial/).

Aloitan suorittamalla `sudo apt-get update` ja sen jälkeen asennan virtualenv paketin komennolla `sudo apt-get install virtualenv`.

Sen jälkeen luodaan env/ kansio ja aktivoidaan se:

![e90da840fad6ad55227567979eb7dab8](https://github.com/user-attachments/assets/bcbf11f3-2933-490f-ae61-ec1d2c5410aa)

Tarkistetaan että kansiopolku on oikein ja luodaan sinne `requirements.txt` tiedosto johon tulee yksi rivi `django`.

![38dcf4f4c07d4c843900981aec399bf0](https://github.com/user-attachments/assets/3406115e-9c5e-4f18-88fc-f4df7a82f13d)

Asennetaan django komennolla `pip install -r requirements.txt` ja varmistetaan että se on asennettu kysymällä versiota.

![5f777e47153e04fa214c39427d4b086c](https://github.com/user-attachments/assets/a59ff073-4c8d-4d65-a645-f193bf760685)

Aloitin uuden projektin nimeltä `krideoy` ja käynnistin sen ajamalla komennon `./manage.py runserver`

![82f767d30a71103882dbe23bf008c6ff](https://github.com/user-attachments/assets/798e13da-faa4-4d83-b0f0-b8c8b6b00fd0)

Seuraavaksi kokeilin selaimessa ja totesin että kaikki toimii.

![419addcdd6166d3af9a7ee6354dda577](https://github.com/user-attachments/assets/4ece0c0a-7a2f-4721-a40f-4d86a4e56be5)

Sitten suljin palvelun hetkeksi ja tein migraatiot komennoilla:

![2f1fde32b64db13ab7f8cb454c8acc95](https://github.com/user-attachments/assets/3e043c31-b7d1-4c01-9163-f67d83003335)

Seuraavaksi asensin pwgen paketin komennolla `sudo apt-get install pwgen` ja generoin uuden salasanan komennolla `pwgen -s 20 1`.

Otin salasanan ylös ja loin itselleni pääkäyttäjän komennolla `./manage.py createsuperuser`, jonka jälkeen käynnistin djangon taas ja menin kokeilemaan tekemiäni muutoksia.

Kaikki näyttäisi toimivan, onnistuin kirjautumaan admin näkymään:

![b7b5679794a417c8af105ec1e5ec208a](https://github.com/user-attachments/assets/dc4fa8c6-a07c-49f0-a4d7-f56b4427d6aa)

Sen jälkeen lisäsin itselleni käyttäjän, valitsin tylsän nimen `admin` ja annoin hyvän salasanan.

![c5a1940b4d579a52ac1499313eb9981b](https://github.com/user-attachments/assets/8500993f-5309-4aa1-b929-961fa2ad1bef)

Suljin taas djangon ja lisäsin crm sekä kävin muokkaamassa `settings.py` tiedostoon uuden rivin `crm`.

![660f4bc47b54b44383814da0020bae99](https://github.com/user-attachments/assets/db53a828-bda9-4f19-a097-cf9a86862e54)

![59bd379b1718e6b698c62ff1910205c3](https://github.com/user-attachments/assets/e45c8aff-695a-4e55-a712-a1488cc2c81e)

Sen jälkeen lisätään tietokantaan uusi luokka nimeltä `Customer`.

Ensin syötin `nano crm/models.py` johon lisäsin uuden luokan:

![9f9d55da7b9155f16f5bc4c51fb4bea3](https://github.com/user-attachments/assets/9ca8df22-1398-45e3-a104-ad773ad292d1)

Suoritin migraatiot uudestaan:

![d50900951bd52acdd92fd2f6db634e08](https://github.com/user-attachments/assets/a7ca8b3b-34d8-422e-9f84-dffae2912427)

Sen jälkeen lisätään vielä `admin.py` tiedostoon muutama rivi koodia, tiedosto aukesi komennolla `nano crm/admin.py`.

![788f7a94b52f9e4881eb46ff06d875f9](https://github.com/user-attachments/assets/0da0d515-0efa-4249-b21b-989beb3663e1)

Sen jälkeen käynnistin taas djangon ja kävin kokeilemassa tekemiäni muutoksia, uusi luokka näkyy!

![08199df15fd05fa62315083448882853](https://github.com/user-attachments/assets/5073832f-5eee-4928-b1d1-83bf1909acee)

Loin itselleni muutamat asiakkaat:

![2bad1de60c220000dde577cffc4f581f](https://github.com/user-attachments/assets/94269613-b925-43f9-8bcf-27fa3553a982)

Sen jälkeen kävin vielä muokkaamassa `nano crm/models.py` jotta nimet näkyisivät kunnolla.

![51fc099ecf26e5fdd4424fbe8da1355c](https://github.com/user-attachments/assets/c413eaa0-6cfd-4236-bbaa-9ea5ea3f26c1)

Käynnistin djangon viimeisen kerran ja varmistin että kaikki toimii:

![2facbe03fb274c5af7e89c7916af92b3](https://github.com/user-attachments/assets/0f14120b-de18-488e-8e50-f06c5a960cba)

Kaikki näyttäisi toimivan ja olen tyytyväinen lopputulokseen!

## Tehtävä b

Suoritan tämän tehtävän Teron luomalla ohjeella: ["Deploy Django 4"](https://terokarvinen.com/2022/deploy-django/).

Aloitin tehtävän luomalla `publicwsgi/krideoy/static/` hakemiston ja sijoittamalla sinne `index.html` tiedoston:

![71e38fd2bb73cabc21fb5aef9281e470](https://github.com/user-attachments/assets/f4c479fe-792a-46b2-a73d-cbd216cd6a11)

Sen jälkeen kävin luomassa sille virtualhost .conf tiedoston:

![f5f4d1d0efa36ad2f5e3604d3f0645d2](https://github.com/user-attachments/assets/3e732d16-4bf2-466b-950c-cd545e805940)

Laitoin sen päälle komennolla `sudo a2ensite krideoy.conf` ja käynnistin palvelimen uudestaan komennolla `sudo systemctl restart apache2`.

Tämän jälkeen kävin vielä NameCheapissä lisäämässä uuden alidomainin eli sivusto tulisi näkyviin `krideoy.kridesav.me`.

Odottelin hetken kunnes se lähti toimimaan:

![1c05b4665667bdc387d93c62dad3f55c](https://github.com/user-attachments/assets/06453aa0-b523-436d-8f96-f513bed409f3)

Olin jo asentanut virtualenvin aikasemmalta kokeilulta joten sen kohdan jätän nyt väliin omassa toiminnassani mutta sen saa helposti komennolla `sudo apt-get install virtualenv`.

Seuravaaksi loin itsellen virtualenv tilan publicwsgi kansioon:

![ef151dc987519199112083005df4b877](https://github.com/user-attachments/assets/b1436ac1-d281-4a3c-9c1c-1f3fc6d497b8)

Viimeisellä rivillä näkyy myös kun otan virtualenvin käyttöön.

Sitten loin taas `requirements.txt` sekä lisäsin sinne yhden rivin: `django`. Asensin djangon ajamalla `pip install -r requirements.txt`.

![a9942ac23c744c6df2d5902132d7b6e5](https://github.com/user-attachments/assets/e3c92432-619b-45d4-8a10-af873cc5a31f)

Lopuksi varmistin vielä version viimeisellä rivillä.

Seuraavaksi kopion testiversion minkä tein edellisessä tehtävässä tänne, eli krideoy kansion edellisestä tehtävästä `publicwsgi` kansioon.

Sen jälkeen kävin tekemässä muutoksia virtualhost tiedostossa:

![0f8eeb2b6ef285cec98b5d51b626478e](https://github.com/user-attachments/assets/5197e0c3-db91-4d55-ad1f-ad699be1c710)

Seuraavaksi asensin `sudo apt-get -y install libapache2-mod-wsgi-py3`.

Tässä kohtaa törmäsin muuriin enkä saanut ohjelmaa toimimaan pitkään aikaan, logeissa näkyy virhe:

<img width="729" alt="Screenshot 2024-10-01 at 14 21 53" src="https://github.com/user-attachments/assets/7a77101c-8faf-4c7b-8dd5-ad2c3a2c152c">

Tarkastin kaikkia mahdollisia kirjoitus virheitä yms. Kunnes parin tunnin jälkeen tajusin että virtualhost .conf tiedostossa pythonin path versio on vanhentunut, koska minulla on python 3.11 ja ohjeessa oli 3.9. Vaihdoin sen ja käynnistin apachen uudestaan taas `sudo systemctl restart apache2`.

Nyt sain sivun vihdoin toimimaan!

<img width="729" alt="Screenshot 2024-10-01 at 14 25 39" src="https://github.com/user-attachments/assets/05169f0f-fb8e-467d-98b6-8862bd1207e0">

Seuraavaksi lähdin muokkaamaan asetuksia komennolla `nano krideoy/settings.py` ja sieltä muutin nämä rivit:

<img width="472" alt="Screenshot 2024-10-01 at 14 30 05" src="https://github.com/user-attachments/assets/1ecc179d-184a-4f33-9a0b-86ec11924472">

Käynnistin apachen uudelleen tutulla komennolla `sudo systemctl restart apache2` ja nyt sivu näytti vain `Not found`, eli jotain sain tehtyä!

Seuraavaksi kokeilin `krideoy.kridesav.me/admin`, sivu toimii mutta teemat puuttuu:

<img width="425" alt="Screenshot 2024-10-01 at 14 34 18" src="https://github.com/user-attachments/assets/2e1b9b09-4fab-4c13-9e8d-f0d55df1d7e0">

Seuraavaksi avasin taas asetukset komennolla `nano krideoy/settings.py`, lisäsin ensimmäiselle riville `import os` ja `STATIC_URL` jälkeen vielä toisen rivin:

<img width="364" alt="Screenshot 2024-10-01 at 14 37 45" src="https://github.com/user-attachments/assets/c60403a2-e8f9-46b8-b273-8876f875173d">

Seuraavaksi ajoin komennon `./manage.py collectstatic`:

<img width="528" alt="Screenshot 2024-10-01 at 14 48 19" src="https://github.com/user-attachments/assets/612d91ed-7366-4c85-aabd-83284086cf74">

Teemat toimii!

<img width="627" alt="Screenshot 2024-10-01 at 14 49 32" src="https://github.com/user-attachments/assets/c622d5bf-1118-4007-ada8-55acd5ef6ff0">

Edellisessä tehtävässä tekemäni tiedot on tallella ja pystyin kirjautumaan:

<img width="462" alt="Screenshot 2024-10-01 at 14 51 52" src="https://github.com/user-attachments/assets/9f7bb370-07c7-487f-abda-1c7f83df22d6">

Täten hyväksyin tehtävän tehdyksi!

Tehtävissä meni yhteensä niin paljon aikaa että jätän extra tehtävät tällä kertaa väliin.






